src/brain.coffee

```diff
@@ -379,61 +374,62 @@ class Brain
       args = []
 
       # get arguments
-      while true
-        m = argsRe.exec(text)
-        if not m
-          break
+      while m = argsRe.exec(text)
         args.push(m[1])
 
-
-      matches[match[1]] =
+      matches.push
         text: text
         obj: subroutineName
         args: args
+        start: match.index
+        length: match[0].length
 
     # go through all the object calls and run functions
-    for k,data of matches
+    for data in matches
       output = ""
       if @master._objlangs[data.obj]
         # We do. Do we have a handler for it?
         lang = @master._objlangs[data.obj]
         if @master._handlers[lang]
           # We do.
-          output = @master._handlers[lang].call(@master, data.obj, data.args, scope, hooks)
+          data.output = @master._handlers[lang].call(@master, data.obj, data.args, scope, hooks)
         else
-          output = "[ERR: No Object Handler]"
+          data.output = "[ERR: No Object Handler]"
       else
-        output = @master.errors.objectNotFound
+        data.output = @master.errors.objectNotFound
 
+      # if we get a promise back and we are not in the async mode,
+      # leave an error message to suggest using an async version of rs
+      # otherwise, keep promises tucked into a list where we can check on
+      # them later
```

Comment:
The logic from here on down is a little tortured due to the multiple async checks - it might make sense to add _resolveCallTags and _resolveCallTagsAsync and organize the rest accordingly.